File: fuzzy.c
#include <stdio.h>

// Fungsi untuk menghitung derajat keanggotaan fuzzy
// rumus fuzzy
float fuzzyMember(float value, float a, float b) {
    // if (value <= a) return 0.0; // batas bawah anggota
    // if (value >= b) return 1.0; // batas atas anggota
    // return (value - a) / (b - a); // menghitung nilai linear keanggotaan
    return (value <= a) ? 0.0 : (value >= b) ? 1.0 : (value - a) / (b - a);
}

// fungsi AC mode auto
// Operasi utama
float autoMode(float level) {
    float low = fuzzyMember(level, 0, 20); // batas rentang anggota
    float medium = fuzzyMember(level, 18, 28);
    float high = fuzzyMember(level, 26, 45);

    printf("Suhu Ruangan: %.2f\n", level);
    printf("Derajat keanggotaan (Low): %.2f\n", low);
    printf("Derajat keanggotaan (Medium): %.2f\n", medium);
    printf("Derajat keanggotaan (High): %.2f\n", high);

    // if (high > 0.5) {
    //     printf("Status AC: Suhu AC diatur ke 18 derajat Celcius\n");
    // } else if (medium > 0.5) {
    //     printf("Status AC: Suhu AC diatur ke 23 derajat celcius\n");
    // } else {
    //     printf("Status AC: AC dalam posisi menutup\n");
    // }

    const char* statusAc =  (high > 0.5) ? "Status AC: Suhu AC diatur ke 18 derajat Celcius\n" : (medium > 0.5) ? "Status AC: Suhu AC diatur ke 23 derajat celcius\n" : "Status AC: AC dalam posisi menutup\n";
    printf("%s\n", statusAc);
}

int main() {
    float suhuRuang;

    printf("Masukkan suhu ruangan (0-45): ");
    scanf("%f", &suhuRuang);

    autoMode(suhuRuang);
    
    return 0;
}

File: fuzzy.js
// tolong install node.js terlebih dahulu
// gunakan "node fuzzy.js" di terminal untuk menjalankan program

// Mengimpor modul readline untuk input user
const readline = require("readline");

// Membuat antarmuka untuk input data dari user
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

// Fungsi untuk menghitung nilai linear keanggotaan dengan batas atas 1.0 dan batas bawah 0.0
const fuzzyMembership = (value, a, b) =>
  value <= a ? 0.0 : value >= b ? 1.0 : (value - a) / (b - a);

// Fungsi untuk menenetukan aksi ac dalam mode auto
function autoModeAC(degree) {
  const low = fuzzyMembership(degree, 0, 30); // batas rentang anggota
  const medium = fuzzyMembership(degree, 20, 70);
  const high = fuzzyMembership(degree, 50, 100);

  console.log(`Suhu ruangan: ${degree.toFixed(2)}`);
  console.log(`Derajat keanggotaan (Low): ${low.toFixed(2)}`);
  console.log(`Derajat keanggotaan (Medium): ${medium.toFixed(2)}`);
  console.log(`Derajat keanggotaan (High): ${high.toFixed(2)}`);

  const statusAC =
    high > 0.5
      ? "Status AC: Suhu AC diatur ke 18 derajat Celcius\n"
      : medium > 0.5
      ? "Status AC: Suhu AC diatur ke 23 derajat celcius\n"
      : "Status AC: AC dalam posisi menutup\n";
  console.log(statusAC);
}

// Fungsi utama untuk meminta input pengguna
rl.question("Masukkan suhu ruang (0-45): ", (input) => {
  const suhuRuang = parseFloat(input);

  // cek input hanya dari 0 - 45
  if (!isNaN(suhuRuang) && suhuRuang >= 0 && suhuRuang <= 100) {
    // output
    autoModeAC(suhuRuang);
  } else {
    console.log("Input tidak valid. Silakan masukkan angka antara 0 dan 45.");
  }
  rl.close(); // Menutup antarmuka readline setelah selesai
});
